overlord/restart/restart.go

```diff
@@ -362,3 +362,126 @@ func ReplaceBootID(st *state.State, bootID string) {
 	rm.bootID = bootID
 }
+
+// markTaskForRestart sets certain properties on a task to mark it for system restart.
+// The status argument is the status that the task will have after the system restart.
+func markTaskForRestart(t *state.Task, status state.Status) {
+	// Preserve previous restart behaviour for classic in the undo cases.
+	if release.OnClassic && (status == state.UndoStatus || status == state.UndoneStatus) {
+		t.SetStatus(status)
+		t.Logf("Skipped automatic system restart on classic system when undoing changes back to previous state")
+		return
+	}
+
+	rm := restartManager(t.State(), "internal error: cannot request a restart before RestartManager initialization")
+
+	// store current boot id to be able to check later if we have rebooted or not
+	t.Set("wait-for-system-restart-from-boot-id", rm.bootID)
+	t.SetToWait(status)
+	setWaitForSystemRestart(t.Change())
+
+	t.Logf("Task set to wait until a system restart allows to continue")
+}
+
+func changeRestartContext(chg *state.Change) (*RestartContext, error) {
+	var rt RestartContext
+	if chg == nil {
+		return nil, fmt.Errorf("task is not bound to any change")
+	}
+
+	if err := chg.Get("restart-ctx", &rt); err != nil {
+		if errors.Is(err, &state.NoStateError{}) {
+			return &RestartContext{}, nil
+		}
+		return nil, err
+	}
+	return &rt, nil
+}
+
+// FinishTaskWithRestart2 either schedules a restart for the given task or it
+// does an immediate restart of the snapd daemon, depending on the type of restart
+// provided.
+// For SystemRestart and friends, the restart is scheduled and postponed until the
+// change has run out of tasks to do, and is then performed in processRestartForChange.
+// For tasks that request restarts as a part of their undo, any tasks that previously scheduled
+// restarts as a part of their 'do' will be unscheduled.
+// XXX: will replace FinishTaskWithRestart
+func FinishTaskWithRestart2(t *state.Task, snapName string, status state.Status, restartType RestartType, rebootInfo *boot.RebootInfo) error {
+	switch restartType {
+	case RestartSystem, RestartSystemNow, RestartSystemHaltNow, RestartSystemPoweroffNow:
+		break
+	default:
+		t.SetStatus(status)
+		Request(t.State(), restartType, rebootInfo)
+		return nil
+	}
+
+	chg := t.Change()
+	rt, err := changeRestartContext(chg)
+	if err != nil {
+		return err
+	}
+
+	if snapName == "" {
+		snapName = "snapd"
+	}
+	rt.setParameters(snapName, restartType, rebootInfo)
+
+	// Either invoked with undone or done as the final status. We don't expect
+	// other uses than tasks that end their Doing/Undoing to call this. Let's
+	// only allow these for now as nothing tests with anything else.
+	switch status {
+	case state.DoneStatus, state.UndoneStatus:
+		markTaskForRestart(t, status)
+	default:
+		return fmt.Errorf("unexpected task status when requesting system restart for task: %s", status)
+	}
+
+	chg.Set("restart-ctx", rt)
```

Comment:
pending-system-restart then here?

Reply:
I think this is still a good idea, especially once we rename the struct

Reply:
renamed to "restart-parameters" to reflect the renamed structure now

Reply:
I still prefer my proposal, it's more consistent with the other Change fields used by restart logic, it contains pending as this data will exist on the change until the pending restart is turned into an actual request

Reply:
Sorry I misunderstood that one for the data identifier - would you prefer `pending-system-restart` or `pending-restart-parameters`?

Reply:
I think "pending-system-restart" is fine, otherwise it would have to be "pending-system-restart-parameters" which seems a bit too much, also would complicate adding more fields that would make it less "parameters"-only if we need to

Reply:
renamed this, thanks!